### PR TITLE
Save Mesh as OBJ

### PR DESCRIPTION
**Pull request by amngupta**
Overview
---
Save SpatialUnderstanding Mesh as an OBJ fie in the HoloLens. 

Steps:
1. Add RoomSaver component to scene
2. Initialize the fileName and anchorStoreName fields
3. On SpatialUnderstanding.ScanStates.Done, call RoomSaver.SaveRoom() method

Changes
---
- Fixes: #188
- Fixes: #984

### CODE REVIEW

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    public GameObject surfaceObject;            // prefab for surface mesh objects
+    public string fileName;                     // name of file used to store mesh
+    public string anchorStoreName;              // name of world anchor for room
+
+    List<Mesh> roomMeshes;                      // list of meshes saved from spatial mapping
+    WorldAnchorStore anchorStore;               // store of world anchors
+
+    // Use this for initialization
+    void Start()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs]: missing private access modifiers in methods and fields.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+using HoloToolkit.Unity;
+using UnityEngine;
+using UnityEngine.VR.WSA.Persistence;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs]: Please make sure this builds for windows standalone as well.
Also, the VR namespace should be XR

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Use this for initialization
+    void Start()
+    {
+        // get instance of WorldAnchorStore
+        WorldAnchorStore.GetAsync(AnchorStoreReady);
+    }
+
+    // Update is called once per frame
+    void Update()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs]: remove unused methods.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+            Debug.Log("Mesh " + filter.mesh.name + ": " + filter.transform.position + "\n--- rotation " + filter.transform.localRotation + "\n--- scale: " + filter.transform.localScale);
+            // add mesh to room meshes for serialization
+            roomMeshes.Add(filter.mesh);
+
+            // save world anchor
+            WorldAnchor attachingAnchor = filter.gameObject.GetComponent<WorldAnchor>();
+            if (attachingAnchor == null)
+            {
+                attachingAnchor = filter.gameObject.AddComponent<WorldAnchor>();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs]: `gamobject.EnsureComponent` would remove the need to check for null.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        string[] anchorIds = anchorStore.GetAllIds();
+        for (int i = 0; i < anchorIds.Length; i++)
+        {
+            if (anchorIds[i].Contains(anchorStoreName))
+            {
+                anchorStore.Delete(anchorIds[i]);
+            }
+        }
+
+        Debug.Log("Old anchors deleted...");
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs]: Do we really need all these debug logs?
Comments should be enough, if you're no longer testing

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
+using HoloToolkit.Unity;
+using UnityEngine;
+using UnityEngine.VR.WSA;
+using UnityEngine.VR.WSA.Persistence;
+using System.Collections.Generic;
+using HoloToolkit.Unity.SpatialMapping;
+
+public class RoomSaver : MonoBehaviour
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs]: Missing namespace

File: Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/SimpleMeshSerializer.cs [hunk 7]
```diff
@@ -52,12 +52,10 @@ public static byte[] Serialize(IEnumerable<Mesh> meshes)
         /// <summary>
         /// Serializes a list of MeshFilter objects into a byte array.
         /// Transforms vertices into world space before writing to the file.
-		/// Optionally transforms the vertices into the space of the supplied secondarySpace
         /// </summary>
         /// <param name="meshes">List of MeshFilter objects to be serialized.</param>
-        /// <param name="secondarySpace">New space to transform the vertices into.</param>
         /// <returns>Binary representation of the Mesh objects.</returns>
-        public static byte[] Serialize(IEnumerable<MeshFilter> meshes, Transform secondarySpace = null)
+        public static byte[] Serialize(IEnumerable<MeshFilter> meshes)
```
Comment (keveleigh) [inline Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/SimpleMeshSerializer.cs]: Why was `secondarySpace` removed? Seems like a useful feature.

File: Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/SimpleMeshSerializer.cs [hunk 8]
```diff
@@ -152,25 +150,25 @@ private static Mesh ReadMesh(BinaryReader reader)
         /// <param name="writer">BinaryWriter representing the data stream.</param>
         /// <param name="vertexCount">Count of vertices in the mesh.</param>
         /// <param name="triangleIndexCount">Count of triangle indices in the mesh.</param>
-        private static void WriteMeshHeader(BinaryWriter writer, int vertexCount, int triangleIndexCount)
+        private static void WriteMeshHeader(BinaryWriter writer, string name, int vertexCount, int triangleIndexCount)
```
Comment (keveleigh) [inline Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/SimpleMeshSerializer.cs]: Please add `name` to the method's param documentation.

File: Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/SimpleMeshSerializer.cs [hunk 9]
```diff
@@ -152,25 +150,25 @@ private static Mesh ReadMesh(BinaryReader reader)
...
         }
-
         /// <summary>
         /// Reads a mesh header from the data stream.
         /// </summary>
         /// <param name="reader">BinaryReader representing the data stream.</param>
         /// <param name="vertexCount">Count of vertices in the mesh.</param>
         /// <param name="triangleIndexCount">Count of triangle indices in the mesh.</param>
-        private static void ReadMeshHeader(BinaryReader reader, out int vertexCount, out int triangleIndexCount)
+        private static void ReadMeshHeader(BinaryReader reader, out string name, out int vertexCount, out int triangleIndexCount)
```
Comment (keveleigh) [inline Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/SimpleMeshSerializer.cs]: Please add `name` to the method's param documentation.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            WorldAnchor attachingAnchor = filter.gameObject.GetComponent<WorldAnchor>();
+            if (attachingAnchor == null)
+            {
+                attachingAnchor = filter.gameObject.AddComponent<WorldAnchor>();
+                Debug.Log("" + filter.mesh.name + ": Using new anchor...");
+            }
+            else
+            {
+                Debug.Log("" + filter.mesh.name + ": Deleting existing anchor...");
+                DestroyImmediate(attachingAnchor);
```
[the remark this answers is not in the document]
Comment (NeerajW) [reply Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs]: https://docs.unity3d.com/ScriptReference/Object.DestroyImmediate.html

Destroys the object obj immediately. You are strongly recommended to use Destroy instead.

**Comment by StephenHodgson** [comment]
@amngupta you still have a few file conflicts

**Comment by DDReaper** [comment]
Should this be re-targeted at Stabalisation @StephenHodgson ?

**Comment by StephenHodgson** [comment]
@amngupta do you need this in the next release? Or is having it go into the dev branch okay with you?

**Comment by amngupta** [comment]
@StephenHodgson can leave on dev branch for now. 

I would like to make some examples/tests as well to showcase the feature!
Help from other developers will be appreciated 😄

**Comment by tontonwawa** [comment]
Hi, I am currently trying to run your code for Hololens and I cannot build with roomloader.cs
There is an error about the GameManager component which is not recognized (line 49 of your file).
Should I use a library that is not stated in the file ?

**Comment by amngupta** [comment]
@tontonwawa you can refer to #188 for details. 

It explains the code very well.

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    {
+
+        public string fileName;             // name of file to store meshes
+        public string anchorStoreName;      // name of world anchor to store for room
+
+        private List<MeshFilter> roomMeshFilters;
+        private WorldAnchorStore anchorStore;
+        private int meshCount = 0;
+
+        // Use this for initialization
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs]: please remove comment

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs [hunk 12]
```diff
@@ -0,0 +1,113 @@
+using HoloToolkit.Unity;
+using UnityEngine;
+using System.Collections.Generic;
+using HoloToolkit.Unity.SpatialMapping;
+using UnityEngine.XR.WSA.Persistence;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomSaver.cs]: breaks cross platform support

File: Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                GameObject obj = Instantiate(surfaceObject) as GameObject;
+                obj.GetComponent<MeshFilter>().mesh = surface;
+                obj.GetComponent<MeshCollider>().sharedMesh = surface;
+
+                if (!anchorStore.Load(surface.name, obj))
+                    Debug.Log("WorldAnchor load failed...");
+
+            }
+
+            GameObject.Find("GameManager").GetComponent<GameManager>().RoomLoaded();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/SpatialProcessing/RoomLoader.cs]: We can probably set a reference to this in the inspector

File: Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/MeshSaver.cs [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+                            {
+                                StorageFolder folder = await StorageFolder.GetFolderFromPathAsync(folderName);
+                                StorageFile file = await folder.CreateFileAsync(fileName, CreationCollisionOption.ReplaceExisting);
+                                IRandomAccessStream randomAccessStream = await file.OpenAsync(FileAccessMode.ReadWrite);
+                                stream = randomAccessStream.AsStreamForWrite();
+                            });
+            task.Wait();
+            task.Result.Wait();
+#else
+            stream = new FileStream(Path.Combine(folderName, fileName), FileMode.Create, FileAccess.Write);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/MeshSaver.cs]: Should we make this an enumeration if this blocks the main thread?
If we don't care, then why use async for WSA?

File: Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/MeshSaver.cs [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+            get
+            {
+#if !UNITY_EDITOR && UNITY_WSA
+                return ApplicationData.Current.RoamingFolder.Path;
+#else
+                return Application.persistentDataPath;
+#endif
+            }
+        }
+        /// <summary>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/SpatialMapping/Scripts/RemoteMapping/MeshSaver.cs]: needs an extra line

**Comment by SimonDarksideJ** [comment]
Any update on this @amngupta ?
Given the minor changes here, this might be a better fit against the latest Master patch release.
Could you create a new PR against the "Patch_4" branch with these changes and fixes.

**Comment by StephenHodgson** [comment]
Closed due to inactivity. Please reopen when you get time to work on this.

**Pull request closed without merging**